subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildRegistry.java

```diff
@@ -240,6 +247,36 @@ private BuildIdentifier idFor(String buildName) {
     }
 
     private Path assignPath(BuildState owner, String name, File dir) {
+        // Get the closest ancestor build of the build directory which we are currently adding
+        Optional<Map.Entry<File, IncludedBuildState>> parentBuild = includedBuildsByRootDir.entrySet().stream()
```

Comment:
I think we potentially have an issue when a nested build is included by some other build before its ancestor is.

This case should work, but we should add some test coverage for it:

1. root includes `build-a`, `build-b` and `nested`, in that order
2. `build-a` includes `build-b/nested`

I think this case will be broken. This might be fine but we might want to check the error message and see if can be improved:

1. root includes `build-a` and `nested`
2. `build-a` includes `build-b/nested`
3. `nested` includes `build-b`

Reply:
I added the two test cases. It seems like the first case fails as well. Is it possible that `includeBuild` works depth-first?

Reply:
Yes, maybe it does. I think we can live with these breakages for now and if someone raises an issue for them we can figure out a way to fix them or at least improve the error message.